src/Adapter/MSTest.CoreAdapter/Helpers/ReflectHelper.cs

```diff
@@ -365,19 +367,20 @@ internal bool IsDoNotParallelizeSet(Assembly assembly)
         /// Gets custom attributes at the class and assembly for a method.
         /// </summary>
         /// <param name="attributeProvider">Method Info or Member Info or a Type</param>
+        /// <param name="owningType">The type that owns <paramref name="attributeProvider"/>.</param>
         /// <param name="type"> What type of CustomAttribute you need. For instance: TestCategory, Owner etc.,</param>
         /// <returns>The categories of the specified type on the method. </returns>
-        internal IEnumerable<object> GetCustomAttributesRecursively(MemberInfo attributeProvider, Type type)
+        internal IEnumerable<object> GetCustomAttributesRecursively(MemberInfo attributeProvider, Type owningType, Type type)
         {
             var categories = this.GetCustomAttributes(attributeProvider, typeof(TestCategoryBaseAttribute));
             if (categories != null)
             {
-                categories = categories.Concat(this.GetCustomAttributes(attributeProvider.DeclaringType.GetTypeInfo(), typeof(TestCategoryBaseAttribute))).ToArray();
+                categories = categories.Concat(this.GetCustomAttributes(owningType.GetTypeInfo(), typeof(TestCategoryBaseAttribute))).ToArray();
             }
 
             if (categories != null)
             {
-                categories = categories.Concat(this.GetCustomAttributeForAssembly(attributeProvider, typeof(TestCategoryBaseAttribute))).ToArray();
+                categories = categories.Concat(this.GetCustomAttributeForAssembly(owningType.GetTypeInfo().Assembly, typeof(TestCategoryBaseAttribute))).ToArray();
```

Comment:
https://docs.microsoft.com/en-us/dotnet/api/system.reflection.memberinfo.declaringtype?view=netframework-4.7.2#System_Reflection_MemberInfo_DeclaringType

It seems the situation you were facing was only because the test method wasn't overridden. From the 4 scenarios in the link, one of the scenarios is this issues particular scenario. Just to be sure can you provide the behavior for the rest of the three scenarios.

I know this seems far fetched, but this could lead to a big breaking change so I just want to make sure :)

Reply:
Good idea. I took the example class hierarchy from that link and added test attributes to get this: https://gist.github.com/spanglerco/fdb04b4a6bab1be2963f6063c937a587

It actually exposes an interesting corner case for classes that hide test methods (not that I can immediately think of a use case for that). Here's the result of using vstest.console to run the tests filtered by one category at a time with 1.4.0:

Category | Test Class Per Execution | Test Method Executed | Total
:--------: | ------------------------ | ----------------------- | :---:
A | <ul><li>A</li><li>B</li><li>B</li><li>C</li><li>D</li></ul> | <ul><li>A.Test</li><li>B.Test</li><li>B.Test</li><li>A.Test</li><li>D.Test</li></ul> | 5
B | <ul><li>B</li></ul> | <ul><li>B.Test</li></ul> | 1
C | -- | -- | 0
D | <ul><li>D</li></ul> | <ul><li>D.Test</li></ul> | 1

The interesting thing is that test category A gets applied to B's `new Test()`, but instead of running both `A.Test()` with an instance of B and `B.Test()`, it runs `B.Test()` twice. But test category B only applies to `B.Test()`, so it runs once for that category. Test category C isn't applied to `A.Test()`, which is my original scenario.

Here's the results with this change:

Category | Test Class Per Execution | Test Method Executed | Total
:--------: | ------------------------ | ----------------------- | :---:
A | <ul><li>A</li><li>B</li><li>B</li><li>C</li><li>D</li></ul> | <ul><li>A.Test</li><li>B.Test</li><li>B.Test</li><li>A.Test</li><li>D.Test</li></ul> | 5
B | <ul><li>B</li><li>B</li></ul> | <ul><li>B.Test</li><li>B.Test</li></ul> | 2
C | <ul><li>C</li></ul> | <ul><li>A.Test</li></ul> | 1
D | <ul><li>D</li></ul> | <ul><li>D.Test</li></ul> | 1

A and D remain the same, but now that the test category of the subclass is applied to inherited methods, test category C includes `A.Test()` on an instance of C. A side-effect is that test category B then includes a second run of `B.Test()` like what happens in test category A.

The reason the duplicate execution happens is because the discoverer only sends the `TestMethod.FullClassName` and not the `TestMethod.DeclaringClassFullName`, so when the executor reacquires the MethodInfo, it does so on B instead of A, making it ambiguous. It simply chooses the first result, which is the `B.Test()` one. Assuming that it's considered backwards compatible to add new property values to the `TestCase`, it seems this could be fixed by sending the declaring class name as an optional property value and the executor can use that to find the correct MethodInfo.

Reply:
@spanglerco, this is really good stuff. The changes look good but can we come up with a solution to avoid running tests multiple times for *overloaded* methods?

Since this is a case for overloading, for test category B the expectation would be that `B.Test` is run only once and since `A.Test` is being overloaded it should not run when running tests for category `B`.

Reply:
> since A.Test is being overloaded it should not run when running tests for category B

@karanjitsingh I should have read this part more clearly to start with. I have a working implementation that resolves the duplicate execution, but I would have thought the desired behavior would be that test category B would run both `B.Test` and `A.Test` on an instance of B since B has both of those methods defined at the same time (`B.Test` is hiding `A.Test`, but `A.Test` is still there):

Category | Test Class Per Execution | Test Method Executed | Total
:--------: | ------------------------ | ----------------------- | :---:
A | <ul><li>A</li><li>B</li><li>B</li><li>C</li><li>D</li></ul> | <ul><li>A.Test</li><li>B.Test</li><li>A.Test</li><li>A.Test</li><li>D.Test</li></ul> | 5
B | <ul><li>B</li><li>B</li></ul> | <ul><li>B.Test</li><li>A.Test</li></ul> | 2
C | <ul><li>C</li></ul> | <ul><li>A.Test</li></ul> | 1
D | <ul><li>D</li></ul> | <ul><li>D.Test</li></ul> | 1

Reply:
@spanglerco, in the case of this type of overload, the language specifies `A.Test` will be run if the object is of type A or the reference for object B is of type A, there is no right way here in this case but I would want to follow the convention that the language follows.

Reply:
@spanglerco, any update on this?

Reply:
@karanjitsingh, sorry for the delay, I haven't had a lot of time to look into this lately. I was able to tweak TypeEnumerator to remove the duplicate methods with these results:

Category | Test Class Per Execution | Test Method Executed | Total
:--------: | ------------------------ | ----------------------- | :---:
A | <ul><li>A</li><li>B</li><li>C</li><li>D</li></ul> | <ul><li>A.Test</li><li>B.Test</li><li>A.Test</li><li>D.Test</li></ul> | 4
B | <ul><li>B</li></ul> | <ul><li>B.Test</li></ul> | 1
C | <ul><li>C</li></ul> | <ul><li>A.Test</li></ul> | 1
D | <ul><li>D</li></ul> | <ul><li>D.Test</li></ul> | 1